Amazon.IonDotnet/Internals/Binary/RawBinaryReader.cs

```diff
@@ -585,14 +585,45 @@ protected Timestamp ReadTimeStamp(int length)
         /// <param name="isNegative">Sign of the value.</param>
         protected BigInteger ReadBigInteger(int length, bool isNegative)
         {
-            // TODO: Improve this
             if (length == 0)
             {
                 return BigInteger.Zero;
             }
 
             var bytes = new byte[length];
             this.ReadAll(bytes, length);
+
+            return this.ToBigInteger(bytes, isNegative);
+        }
+
+        /// <summary>
+        /// Convert <paramref name="bytes"/> into a <see cref="BigInteger"/>.
+        /// </summary>
+        /// <returns>The big integer.</returns>
+        /// <param name="bytes">The bytes corresponding to the UInt magnitude of the int.</param>
+        /// <param name="isNegative">Sign of the int.</param>
+        protected BigInteger ToBigInteger(byte[] bytes, bool isNegative)
+        {
+            if ((bytes[0] & 0x80) > 0)
+            {
+                // The .NET framework currently supported by this library (netstandard2.0)
+                // does not provide a <code>BigInteger(ReadOnlySpan<Byte>, Boolean, Boolean)</code>
+                // (https://docs.microsoft.com/en-us/dotnet/api/system.numerics.biginteger.-ctor?view=netcore-3.1#System_Numerics_BigInteger__ctor_System_ReadOnlySpan_System_Byte__System_Boolean_System_Boolean_)
+                // constructor, so we currently use <code>BigInteger(Byte[])</code>
+                // (https://docs.microsoft.com/en-us/dotnet/api/system.numerics.biginteger.-ctor?view=netcore-3.1#System_Numerics_BigInteger__ctor_System_Byte___)
+                // instead.  This constructor expects the bytes to include a sign bit, whereas
+                // the Ion binary representation is unsigned.
+                //
+                // In cases where the high-order bit of the UInt will be interpreted as a negative sign,
+                // this block prepends a 0x00 byte to force <code>BigInteger(Byte[])</code> to interpret
+                // the bytes as a positive value.  Then we simply call <code>BigInteger.Negate()</code>
+                // if appropriate.
```

Comment:
It looks like this affects large positive ints with the high-order bit set too, correct? Because the highest bit of data would be interpreted instead as the sign bit.

I bring this up because it looks like the added tests only cover large negative ints.

Reply:
Correct, and thank you for calling this out.  I've added test cases for positive ints.